packages/celltags-extension/src/celltag.tsx

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) Jupyter Development Team.
+ * Distributed under the terms of the Modified BSD License.
+ */
+
+import React from 'react';
+import { FieldProps } from '@rjsf/utils';
+import { INotebookTracker } from '@jupyterlab/notebook';
+import { addIcon, checkIcon, LabIcon } from '@jupyterlab/ui-components';
+import { reduce } from '@lumino/algorithm';
+import {
+  ITranslator,
+  nullTranslator,
+  TranslationBundle
+} from '@jupyterlab/translation';
+
+export class CellTagField {
+  constructor(tracker: INotebookTracker, translator?: ITranslator) {
+    this._tracker = tracker;
+    this._translator = translator || nullTranslator;
+    this._trans = this._translator.load('jupyterlab');
+    this._editing = false;
+  }
+
+  addTag(props: FieldProps, tag: string) {
+    const data = props.formData;
+    if (tag && !data.includes(tag)) {
+      data.push(tag);
+      props.formContext.updateMetadata({ [props.name]: data }, true);
+    }
+  }
+
+  /**
+   * Pull from cell metadata all the tags used in the notebook and update the
+   * stored tag list.
+   */
+  pullTags(): string[] {
+    const notebook = this._tracker?.currentWidget;
+    const cells = notebook?.model?.cells;
+    if (cells === undefined) {
+      return [];
+    }
+    const allTags = reduce(
+      cells,
+      (allTags: string[], cell) => {
+        const tags: string[] = (cell.getMetadata('tags') as string[]) ?? [];
+        return [...allTags, ...tags];
+      },
+      []
+    );
+    return [...new Set(allTags)].filter(tag => tag !== '');
+  }
+
+  private _emptyAddTag(target: HTMLInputElement) {
+    target.value = '';
+    target.style.width = '49px';
+  }
+
+  private _onAddTagKeyDown(
+    props: FieldProps,
+    event: React.KeyboardEvent<HTMLInputElement>
+  ) {
+    const input = event.target as HTMLInputElement;
+
+    if (event.ctrlKey) return;
+
+    if (event.key === 'Enter') {
+      this.addTag(props, input.value);
+    } else if (event.key === 'Escape') {
+      this._emptyAddTag(input);
+    }
+  }
+
+  private _onAddTagFocus(event: React.FocusEvent<HTMLInputElement>) {
+    if (!this._editing) {
+      (event.target as HTMLInputElement).blur();
+    }
+  }
+
+  private _onAddTagBlur(input: HTMLInputElement) {
+    if (this._editing) {
+      this._editing = false;
+      this._emptyAddTag(input);
+    }
+  }
+
+  private _onChange(event: React.ChangeEvent<HTMLInputElement>) {
+    if (!event.target.value) {
+      this._emptyAddTag(event.target);
+    } else {
+      const tmp = document.createElement('span');
+      tmp.className = 'add-tag';
```

Comment:
Should all caps global const instead of inline CSS class string be used here?

Please see `shell.ts` for reference:
https://github.com/jupyterlab/jupyterlab/blob/82e1a74b61a41fbe5d65e8542053b749cdfe6ea0/packages/application/src/shell.ts#L39